pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,596 @@
+                                .getSnapshot().getAborts());
+                        if (indexes.size() != 0) {
+                            persistentWorker.sequenceID.set(indexes.lastEntry().getValue().sequenceID + 1);
+                        }
+                    }
+                    //Read snapshot segment to recover aborts.
+                    ArrayList<CompletableFuture<Void>> completableFutures = new ArrayList<>();
+                    CompletableFuture<Void> openManagedLedgerFuture = new CompletableFuture<>();
+                    AtomicLong invalidIndex = new AtomicLong(0);
+                    AsyncCallbacks.OpenReadOnlyManagedLedgerCallback callback = new AsyncCallbacks
+                            .OpenReadOnlyManagedLedgerCallback() {
+                        @Override
+                        public void openReadOnlyManagedLedgerComplete(ReadOnlyManagedLedgerImpl readOnlyManagedLedger,
+                                                                      Object ctx) {
+                            persistentSnapshotIndexes.getIndexList().forEach(index -> {
+                                CompletableFuture<Void> handleSegmentFuture = new CompletableFuture<>();
+                                completableFutures.add(handleSegmentFuture);
+                                readOnlyManagedLedger.asyncReadEntry(
+                                        new PositionImpl(index.getPersistentPositionLedgerID(),
+                                                index.getPersistentPositionEntryID()),
+                                        new AsyncCallbacks.ReadEntryCallback() {
+                                            @Override
+                                            public void readEntryComplete(Entry entry, Object ctx) {
+                                                //Remove invalid index
+                                                if (entry == null) {
+                                                    indexes.remove(new PositionImpl(
+                                                            index.getMaxReadPositionLedgerID(),
+                                                            index.getMaxReadPositionEntryID()));
+                                                    handleSegmentFuture.complete(null);
+                                                    invalidIndex.getAndIncrement();
+                                                    return;
+                                                }
+                                                handleSnapshotSegmentEntry(entry);
+                                                handleSegmentFuture.complete(null);
+                                            }
+
+                                            @Override
+                                            public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+                                                handleSegmentFuture.completeExceptionally(exception);
```

Comment:
When deleting a Segment, we delete the Segment first and then update the Index. If the delete Segment succeeds, but the updated Index fails, then the error occurs here.

Reply:
If the deleted segment succeeds, but the updated index fails. `invalidIndex` will be larger than 0, and then TB will update the index again when TB recovers.

Reply:
No, it throws an `NonRecoverableLedgerException` exception if the entry does not exist

Reply:
At the beginning of the design, it was considered that `write/delete snapshots` and `update index` were non-atomic operations. The update index operation after the `write snapshot` may fail. Here re-update index is the pudding for `update index` failure after `delete snapshots`. This is an expected behavior.